pallets/order-book/src/lib.rs

```diff
@@ -0,0 +1,639 @@
+			);
+			ensure!(
+				T::ReserveCurrency::can_reserve(
+					&account,
+					T::Fees::fee_value(T::OrderFeeKey::get())
+				),
+				Error::<T>::InsufficientReserveFunds,
+			);
+			<NonceStore<T>>::try_mutate(|n| {
+				*n = n.ensure_add(T::Nonce::one())?;
+				Ok::<_, DispatchError>(())
+			})?;
+			let max_sell_amount = buy_amount.ensure_mul(sell_price_limit)?;
+
+			ensure!(
+				T::TradeableAsset::can_reserve(currency_out, &account, max_sell_amount),
+				Error::<T>::InsufficientAssetFunds,
+			);
+
+			let new_nonce = <NonceStore<T>>::get();
+			let order_id = Self::gen_hash(&account, currency_in, currency_out, new_nonce);
+			let new_order = Order {
+				order_id,
+				placing_account: account.clone(),
+				asset_in_id: currency_in,
+				asset_out_id: currency_out,
+				buy_amount,
+				price: sell_price_limit,
+				initial_buy_amount: buy_amount,
+				min_fullfillment_amount,
+				max_sell_amount,
+			};
+
+			<AssetPairOrders<T>>::try_mutate(currency_in, currency_out, |orders| {
+				orders
+					.try_push(order_id)
+					.map_err(|_| Error::<T>::AssetPairOrdersOverflow)
+			})?;
+
+			T::ReserveCurrency::reserve(&account, T::Fees::fee_value(T::OrderFeeKey::get()))?;
+
+			T::TradeableAsset::reserve(currency_out, &account, max_sell_amount)?;
+
+			<Orders<T>>::insert(order_id, new_order.clone());
+			<UserOrders<T>>::insert(&account, order_id, new_order);
+			Self::deposit_event(Event::OrderCreated {
+				creator_account: account,
+				sell_price_limit,
+				order_id,
+				buy_amount,
+				currency_in,
+				currency_out,
+				min_fullfillment_amount,
+			});
+			Ok(order_id)
+		}
+
+		/// Cancel an existing order.
+		/// Unreserve currency reserved for trade as well storage fee.
+		fn cancel_order(order: Self::OrderId) -> DispatchResult {
+			let order = <Orders<T>>::get(order)?;
+			let account_id = order.placing_account;
+
+			T::ReserveCurrency::unreserve(&account_id, T::Fees::fee_value(T::OrderFeeKey::get()));
+			T::TradeableAsset::unreserve(order.asset_out_id, &account_id, order.max_sell_amount);
+			Self::remove_order(order.order_id)?;
+			Self::deposit_event(Event::OrderCancelled {
+				account: account_id,
+				order_id: order.order_id,
+			});
+
+			Ok(())
+		}
+
+		/// Update an existing order.
+		/// Update outgoing asset currency reserved to match new amount or price
+		/// if either have changed.
+		fn update_order(
+			account: T::AccountId,
+			order_id: Self::OrderId,
+			buy_amount: T::ForeignCurrencyBalance,
+			sell_price_limit: T::ForeignCurrencyBalance,
+			min_fullfillment_amount: T::ForeignCurrencyBalance,
+		) -> DispatchResult {
+			ensure!(
+				buy_amount != T::ForeignCurrencyBalance::zero(),
+				Error::<T>::InvalidBuyAmount
+			);
+			ensure!(
+				sell_price_limit != T::ForeignCurrencyBalance::zero(),
+				Error::<T>::InvalidMinPrice
+			);
+
+			<Orders<T>>::try_mutate_exists(order_id, |maybe_order| -> DispatchResult {
+				let mut order = maybe_order.as_mut().ok_or(Error::<T>::OrderNotFound)?;
```

Comment:
Should this always exist because it's inside of a `try_mutate_exists`?

Reply:
As long as a valid order is submitted it should always exist.

Reply:
But if it was not submitted, will the callback of `try_mutate_exists` not be called?

Reply:
Ah, I meant if a valid order id is passed to the update fn.  If there isn't a matching order then it should return an error for no matching order.